### ISSUE TITLE
Generate real Byron transactions

### ISSUE THREAD

**Issue by mrBliss** [input-output-hk/ouroboros-network#888]
We need a `TxGen` (see #855) instance for `ByronBlockOrEBB ByronConfig`. Currently, we're not generating any transactions in the consensus tests for real Byron blocks (we are for mock blocks).

It would be great if we could reuse the existing generators from [cardano-ledger](https://github.com/input-output-hk/cardano-ledger/).

**Comment by ruhatch** [comment]
@intricate do you think you could take this on? The idea was to do something similar to your UTxO tests in `cardano-ledger`, where you generate and elaborate.

**Comment by mrBliss** [comment]
Notes:

* `cardano-ledger` contains [Hedgehog generators](https://github.com/input-output-hk/cardano-ledger/blob/abef065b4c4f2b7edaebed06729e6bcd7fc803a1/cardano-ledger/test/Test/Cardano/Chain/UTxO/Gen.hs) for the UTxO and transactions, but they simply generate arbitrary values for most fields. The chance that a transaction is valid w.r.t. to some UTxO is close to zero.

  Conclusion: we cannot use these generators to generate real Byron transactions.

* `cardano-ledger-specs` contains [Hedgehog generators](https://github.com/input-output-hk/cardano-ledger/blob/abef065b4c4f2b7edaebed06729e6bcd7fc803a1/cardano-ledger/test/Test/Cardano/Chain/UTxO/Gen.hs) for an *abstract* `UTxO` and *abstract* transactions. Using these generators, we can generate a valid transaction w.r.t. to some UTxO. The generated abstract transaction first has to be *elaborated* into a concrete transaction. Simply put: abstract = `cardano-ledger-specs` types, concrete = `cardano-ledger` types. The `cardano-ledger` repository contains [functions](https://github.com/input-output-hk/cardano-ledger/blob/abef065b4c4f2b7edaebed06729e6bcd7fc803a1/cardano-ledger/test/Test/Cardano/Chain/Elaboration/UTxO.hs) to perform this elaboration.

  In our situation we have a concrete UTxO (current ledger state) for which we want to generate valid and invalid concrete transactions. We have the following steps:
  ```
  concrete UTxO ---> abstract UTxO ---> abstract TX ---> concrete TX
                (1)                (2)              (3)
  ```
  1. *abstract* a concrete UTxO into an abstract UTxO: **missing**
  2. generate an abstract transaction w.r.t. an abstract UTxO: available
  3. elaborate an abstract transaction into a concrete transaction: available

  So step 1 is missing. I'm not sure about the feasibility of it. Moreover, I don't think we, consensus, should be the ones implementing it. The ledger teams seem better positioned to do this (if feasible, of course). So for now I eliminate this route.

* `cardano-ledger` and `cardano-ledger-specs` use Hedgehog for property testing and have Hedgehog generators, whereas we, the consensus team, use QuickCheck for property testing and want QuickCheck generators.

  Fortunately, there exists [an easy way](http://hackage.haskell.org/package/hedgehog-quickcheck-0.1.1/docs/Test-QuickCheck-Hedgehog.html) to create a QuickCheck generator based on a Hedgehog generator. It loses Hedgehog's shrinking, though. But that is something we could potentially live with.

* Actually, the context in which we generate these transactions does not allow us to use a QuickCheck nor Hedgehog `Gen` directly: it is not being run directly inside a property. It is in fact in a `MonadRandom m => ... -> m (GenTx blk)` [context](https://github.com/input-output-hk/ouroboros-network/blob/e3a2401aa324afe1287c89d827a92f2e784a3da0/ouroboros-consensus/test-consensus/Test/Dynamic/TxGen.hs#L30).

  However, I have managed to plug in a Hedgehog generator in this context:

  ```haskell
  ..
  import           Hedgehog (Seed (..), Size (..))
  import           Hedgehog.Internal.Gen (Gen, evalGen)
  import           Hedgehog.Internal.Tree (treeValue)
  ..

  testGenTx _numCoreNodes cfg _ledgerState = runGen genGenTx
    where
      runGen :: forall m a. MonadRandom m => Gen a -> m a
      runGen gen = do
        seed <- genSeed
        case evalGen (Size 30) seed gen of
          Nothing -> runGen gen
          Just ta -> return $ treeValue ta

      genSeed :: forall m. MonadRandom m => m Seed
      genSeed = do
        let mn = toInteger (minBound :: Word64)
            mx = toInteger (maxBound :: Word64)
        a <- fromInteger <$> generateBetween mn mx
        b <- fromInteger <$> generateBetween mn mx
        return $ Seed a (if even b then succ b else b)

      -- Dumb generation of random transactions
      genGenTx :: Gen (GenTx (ByronBlockOrEBB cfg))
      genGenTx = mkByronTx . UTxO.annotateTxAux <$>
        UTxO.genTxAux protocolMagicId

      protocolMagicId = Genesis.gdProtocolMagicId
                      . Genesis.configGenesisData
                      . pbftGenesisConfig
                      . encNodeConfigExt
                      . unWithEBBNodeConfig
                      $ cfg
  ```

In conclusion: I believe (with advice from @dnadales) we may have to bite the bullet and write a(n) (in)valid transaction ourselves that looks at the current (concrete) UTxO.

**Comment by dnadales** [comment]
@mrBliss 

> cardano-ledger contains Hedgehog generators for the UTxO and transactions, but they simply generate arbitrary values for most fields. The chance that a transaction is valid w.r.t. to some UTxO is close to zero.

Note that these aren't the generators we use for conformance testing. The UTxO generators, which generate transactions w.r.t. a valid **abstract** UTxO, are in [`module Ledger.UTxO.Generators`](https://github.com/input-output-hk/cardano-ledger-specs/blob/09ca95901d6b1690f10886e690d1adb74df40923/byron/ledger/executable-spec/src/Ledger/UTxO/Generators.hs#L6).

**Comment by dnadales** [comment]
Oh, it seems you're aware of this :) I should have kept on reading ...

**Comment by mrBliss** [comment]
Closing in favour of #1194.

**Comment by edsko** [comment]
The suggestions from @mrBliss above are incorporated in https://github.com/input-output-hk/ouroboros-network/pull/1503 .
